File: courses/web-scraping-ts/unit03-selectors-and-tables/ex02_table_parse.ts
// ex02_table_parse: HTMLのtableを行×列の2次元データに変換する
// tableはthead(見出し行)とtbody(データ行)、その中のtr(行)・td(セル)という
// 決まった階層構造を持つ。見出しをキー、各行のセルを値にすればArray<Record<string, string>>
// という「行の配列」が作れる。C#で言えばDataTableをList<Dictionary<string,string>>に
// 変換する感覚に近い。

import * as cheerio from "cheerio";
import { fileURLToPath } from "node:url";
import path from "node:path";
import fs from "node:fs";

const _unitDir = path.dirname(fileURLToPath(import.meta.url));
const _courseDir = path.basename(path.dirname(_unitDir)) === ".solutions"
  ? path.dirname(path.dirname(_unitDir))
  : path.dirname(_unitDir);
const DATA_DIR = path.join(_courseDir, "data");

// data/ 以下のファイルをUTF-8で読み込んで文字列として返す
export function readFixture(filename: string): string {
  return fs.readFileSync(path.join(DATA_DIR, filename), "utf-8");
}

// table_stats.html の thead th のテキストを出現順の配列で返す
// 例: ["月", "来店者数", "売上冊数"]
export function extractHeaders(html: string): string[] {
  // TODO: "table thead th" をセレクタで取得しテキストの配列にする
  throw new Error("TODO: 未実装");
}

// table_stats.html の tbody tr をそれぞれ td のテキストの配列に変換し、
// 行の配列(2次元配列)として返す
// 例: [["4月", "320", "145"], ["5月", "410", "198"], ...]
export function extractRows(html: string): string[][] {
  // TODO: "table tbody tr" を1行ずつ.each()で処理し、各行の中の"td"を
  // テキストの配列にして集める(行ごとに.find("td")で子孫のtdを探す)
  throw new Error("TODO: 未実装");
}

// ヘッダー配列と行(文字列配列)の配列を受け取り、Array<Record<string, string>>に変換する
// 例: headers=["月","来店者数"], rows=[["4月","320"]]
//     -> [{ 月: "4月", 来店者数: "320" }]
// ヒント: C#のLINQで zip(headers, row) してDictionaryを作る発想に近い。
// 行の列数がheadersより少ない場合、足りない列は無視してよい(ズレの対処はex03の範囲)
export function zipToRecords(headers: string[], rows: string[][]): Array<Record<string, string>> {
  // TODO: 各行についてheadersとインデックスを対応づけ、{ ヘッダー名: セルの値 } の
  // オブジェクトを作る。行のtoRecordはheaders.forEach((h, i) => ...)などで組み立てられる
  throw new Error("TODO: 未実装");
}

// table_stats.html を丸ごと Array<Record<string, string>> に変換する統合関数
// (extractHeaders・extractRows・zipToRecordsを組み合わせるだけでよい)
export function parseStatsTable(html: string): Array<Record<string, string>> {
  // TODO: 上の3つの関数を組み合わせて1行の配列にまとめる
  throw new Error("TODO: 未実装");
}

File: courses/web-scraping/unit03-selectors-and-tables/ex02_table_parse.py
# ex02_table_parse: HTMLのtableを行×列の2次元データに変換する
# table > thead > tr > th でヘッダ、tbody > tr > td で各行のデータが並ぶ。
# C#で言えばDataTableに読み込んで列名でアクセスできるようにする作業に近い。

from pathlib import Path
from bs4 import BeautifulSoup

_UNIT_DIR = Path(__file__).resolve().parent
_COURSE_DIR = _UNIT_DIR.parents[1] if _UNIT_DIR.parent.name == ".solutions" else _UNIT_DIR.parent
DATA_DIR = _COURSE_DIR / "data"


# data/ 以下のHTMLファイルをBeautifulSoupで読み込んで返す
def load_soup(filename):
    path = DATA_DIR / filename
    return BeautifulSoup(path.read_text(encoding="utf-8"), "html.parser")


# soup(table_stats.html)のthead内のth要素から、見出し文字列のリストを返す
# 例: ["月", "来客数", "売上(円)"]
def get_headers(soup):
    # TODO: theadの中のth要素をすべて選び、テキストのリストにする
    raise NotImplementedError


# soup(table_stats.html)のtbody内の各trについて、td要素のテキストのリストを作り、
# それを1行分のリストとしてまとめた「行のリスト」を返す(2次元リスト)
# 例: [["4月", "320", "486000"], ["5月", "410", "612500"], ...]
def get_rows(soup):
    # TODO: tbody内の各tr行について、td要素のテキストのリストを作り集める
    raise NotImplementedError


# get_headers()とget_rows()の結果を組み合わせ、1行=1辞書のリストに変換する
# 例: [{"月": "4月", "来客数": "320", "売上(円)": "486000"}, ...]
# C#で言えばDataRowを列名キーのDictionaryに変換するイメージ
def rows_to_dicts(headers, rows):
    # TODO: 各行をheadersとzipして辞書化し、リストにまとめる
    raise NotImplementedError
